modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,389 @@
+
+		return $this->name;
+	}
+
+	/**
+	 * Determines whether a job should run for current request or not.
+	 *
+	 * It determines by checking if job is already running or completed.
+	 * It also checks if max number of attempts have been attempted for the job.
+	 * For any such condition, it will return false, else true.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool Flag to tell whether this job should run or not.
+	 */
+	public function should_run() {
+		// If job doesn't exist or completed, return false.
+		if ( ! $this->exists() || $this->is_completed() ) {
+			return false;
+		}
+
+		if ( $this->is_running() ) {
+			return false;
+		}
+
+		/**
+		 * Number of attempts to try to run a job.
+		 *
+		 * Repeated attempts may be required to run a failed job. Default 3 attempts.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param int $attempts Number of attempts allowed for a job to run. Default 3.
+		 */
+		$max_attempts = (int) apply_filters( 'perflab_job_max_attempts_allowed', 3 );
+
+		// If number of attempts have been exhausted, return false.
+		if ( $this->get_attempts() >= $max_attempts ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Sets the status of the current job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $status Status to set for current job.
+	 * @return bool Returns true if status is updated, false otherwise.
+	 */
+	public function set_status( $status ) {
+		$valid_statuses = array(
+			self::JOB_STATUS_COMPLETE,
+			self::JOB_STATUS_FAILED,
+			self::JOB_STATUS_QUEUED,
+			self::JOB_STATUS_RUNNING,
+			self::JOB_STATUS_PARTIAL,
+		);
+
+		if ( in_array( $status, $valid_statuses, true ) ) {
+			update_term_meta( $this->id, self::META_KEY_JOB_STATUS, $status );
+
+			// If job is complete, set the timestamp at which it was completed.
+			if ( self::JOB_STATUS_COMPLETE === $status ) {
+				update_term_meta( $this->id, self::META_KEY_JOB_COMPLETED_AT, time() );
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Records the error for the current process run.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Error $error Error instance.
+	 *
+	 * @todo Implement the error recording activity.
+	 *
+	 * @return void
+	 */
+	public function set_error( WP_Error $error ) {
+		$job_failure_data = $error->get_error_data( 'perflab_job_failure' );
+
+		if ( ! empty( $job_failure_data ) ) {
+			$this->set_status( self::JOB_STATUS_FAILED );
+
+			update_term_meta( $this->id, self::META_KEY_JOB_ERRORS, $job_failure_data );
+			update_term_meta( $this->id, self::META_KEY_JOB_ATTEMPTS, ( $this->get_attempts() + 1 ) );
+		}
+	}
+
+	/**
+	 * Retrieves the job status from its meta information.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job status.
+	 */
+	public function get_status() {
+		return (string) get_term_meta( $this->id, self::META_KEY_JOB_STATUS, true );
+	}
+
+	/**
+	 * Returns the number of attempts executed for a job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int
```

Comment:
Please add a return description, e.g.
```suggestion
	 * @return int Number of times the job has been attempted.
```